File: Roadmap/04 - CADENAS DE CARACTERES/javascript/FabianRpv.js
// Cadena de Caracteres

let cadena = "Hola Mundo ";
console.log(cadena);

// Acceso a caracteres especificos

console.log(cadena[0]); // H

// Subcadenas

console.log(cadena.substring(0, 5)); // Hola
console.log(cadena.slice(-5)); // Mundo

// Longitud

console.log(cadena.length); // 10

// Concatenación

let cadena2 = "desde JavaScript";
console.log(cadena + cadena2); // Hola Mundo desde JavaScript

// Repetición

console.log(cadena.repeat(3)); // Hola Mundo Hola Mundo Hola Mundo

// Recorrido

for (let i = 0; i < cadena.length; i++) {
  console.log(cadena[i]);
}

// Conversion a mayusculas y minusculas

console.log(cadena.toUpperCase()); // HOLA MUNDO
console.log(cadena.toLowerCase()); // hola mundo

// Reemplazo

console.log(cadena.replace("Mundo", "Fabian")); // Hola Fabian

// Division 

console.log(cadena.split(" ")); // ["Hola", "Mundo"]

// Union 

console.log(cadena.split(" ").join(" ")); // Hola Mundo

// Interpolación

console.log(`Hola ${cadena2}`); // Hola desde JavaScript

// Verificacion 

console.log(cadena.includes("Hola")); // true
console.log(cadena.startsWith("Hola")); // true
console.log(cadena.endsWith("Mundo ")); // true

// Eliminacion de espacios en blanco

console.log(cadena.trim()); // Hola Mundo

// Busqueda de indice

console.log(cadena.indexOf("Mundo")); // 5
console.log(cadena.lastIndexOf("o")); // 9

/*
Crea un programa que analice dos palabras diferentes y realice comprobaciones para descubrir si son:
- Palíndromos
- Anagramas
- Isogramas
*/

let palabra1 = "Amor";
let palabra2 = "Roma";

// Palindromo

const palindromo = palabra => {
    let palabraInvertida = palabra.toLowerCase().split("").reverse().join("");
    return palabra.toLowerCase() === palabraInvertida;
}

console.log(`Es palindromo: ${palindromo(palabra1)}`);
console.log(`Es palindromo: ${palindromo(palabra2)}`);

// Anagrama

const anagrama = (palabra1, palabra2) => {
    return palabra1.toLowerCase().split('').sort().join('') === palabra2.toLowerCase().split('').sort().join('');
}

console.log(`Es anagrama: ${anagrama(palabra1, palabra2)}`);

// Isograma

const isograma = palabra => {
    let palabraArray = palabra.toLowerCase().split("");
    let palabraSet = new Set(palabraArray);
    return palabraArray.length === palabraSet.size;
}

console.log(`Es isograma: ${isograma(palabra1)}`);
console.log(`Es isograma: ${isograma(palabra2)}`);

File: Roadmap/04 - CADENAS DE CARACTERES/python/FabianRpv.py
# Operaciones con Cadena de Caracteres

str1 = "Hola"
str2 = "Mundo"

concatenacion = str1 + " " + str2  # Concatenar Strings
print(concatenacion)

repeticion = str1*3
print(repeticion)  # Repetir Strings

longitud = len(str1)  # Ver la longitud de un String
print(longitud)

indexacion = str1[0]  # Acceder a una posicion del String
print(indexacion)

busqueda = "H" in str1  # Verificar si existe un valor en el String
print(busqueda)

inicio = str1.startswith("Hol")  # Verificar si inicia el String con...
fin = str1.endswith("la")  # Verificar si finaliza el String con...
print(inicio)
print(fin)

posicion = str1.index("o")  # Devuelve la posicion si el valor existe en el String
print(posicion)

minusculas = concatenacion.lower()  # Convierte el String a Minusculas
print(minusculas)

mayusculas = concatenacion.upper()  # Convierte el String a Mayusculas
print(mayusculas)

title = concatenacion.title()  # Convierte la primera letra de cada palabra en Mayuscula
print(title)

capitalize = concatenacion.capitalize()  # Convierte la primera letra de la oracion en Mayuscula
print(capitalize)

ordenar =  sorted(str1)  # Ordena la lista segun su valor unicode
print(ordenar)

print("  hola  ".strip())   # Elimina espacios
print("  hola  ".lstrip())  # Elimina espacios al principio
print("  hola  ".rstrip())  # Elimina espacios al final

remplazo = concatenacion.replace("Mundo", "Python")  # Reemplaza valores en los Strings
print(remplazo)

union = ", ".join(["a", "b", "c"]) # Une datos en un String
print(union)

division = union.split(",")  # Dividir caracteres de un String en una lista
print(division)

formateo = "Python"  # Formatear datos en un String
print(f"Hola {formateo}" ) 

alphanumerico = str1.isalnum()  # Valida si la cadena es Alfanumerica
print(alphanumerico)

letras = str1.isalpha()  # Verifica si la cadena solo contiene letras
print(letras)

numeros = str1.isdigit()  # Verifica si la cadena solo contiene numeros
print(numeros)

minuscula = str1.islower()  # Verifica si la cadena solo contiene minusculas
print(minuscula)

mayuscula = str1.isupper()  # Verifica si la cadena solo contiene mayusuculas
print(mayuscula)

espacios = str1.isspace()  # Verifica si la cadena solo contiene espacios
print(espacios)

codificacion = "año"
bytes = codificacion.encode('utf8')  # Codifica los caracteres
codificacion = bytes.decode('utf8')  # Descodifica los caracteres
print(bytes)
print(codificacion)

invertir = str1[::-1]  # Invertir cadena
print(invertir) 




# Ejercicio: 

palabra1 = "mora"
palabra2 = "amora"

def es_palindromo(palabra):
    
    if palabra == palabra[::-1]:
        print(f"La palabra {palabra} es un palindromo")
    
    else:
        print(f"La palabra {palabra} no es un palindromo")
        
   
def es_anagrama(palabra1, palabra2):
       
    if sorted(palabra1) == sorted(palabra2):
        
        print(f"Las palabras {palabra1} y {palabra2} son anagramas")
    
    else: 
        
        print(f"Las palabras {palabra1} y {palabra2} no son anagramas")
                

def es_isograma(palabra):
    
    if len(palabra) == len(set(palabra)):
        
        print(f"La palabra {palabra} es un isograma")
        
    else:
            
        print(f"La palabra {palabra} no es un isograma")

es_palindromo(palabra1)
es_palindromo(palabra2)
es_anagrama(palabra1, palabra2)
es_isograma(palabra1)
es_isograma(palabra2)